### PR TITLE
Api peer removal

### PR DESCRIPTION
**Pull request by WolfWindOW**
Adding script to create data for latency graph, as well as a utility script to parse the logs

### CODE REVIEW

File: evaluation/diagramLogs.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import sys
+from datetime import datetime
+
+
+#Takes two times, of format HH:MM:SS and finds the difference
+def calculateTimeDiff(timeOne, timeTwo):
```
Comment (WolfWindOW) [inline evaluation/diagramLogs.py:6]: Takes two times, normalizes them, and calculates the difference between them

File: evaluation/diagramLogs.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        correctionPartTwo = datetime(2021, 11, 25, 13)
+        correct = correctionPartOne - correctionPartTwo
+        postStripOne.__add__(correct)
+        postStripTwo.__add__(correct)
+    timeDifference = postStripTwo - postStripOne
+    return int(timeDifference.total_seconds())
+
+
+#Used in finding timing diagrams
+if __name__ == '__main__':
```
Comment (WolfWindOW) [inline evaluation/diagramLogs.py:20]: Iterates over each line in each log file given

File: evaluation/diagramLogs.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    resultsSet = {}
+    resultsLink = []
+    for i in range(1, len(sys.argv), 1):
+        linkLines = []
+        #setFlag = False
+        with open(sys.argv[i], "r") as f:
+            fileList = list(f)
+            #prev = None
+            for line in fileList:
+                if line.find("intkey set") != -1:
```
Comment (WolfWindOW) [inline evaluation/diagramLogs.py:30]: For each line that has the words "intkey set" in them, ie, whenever a tx is submitted. Then, the transaction number will be placed into a map as a key , with the time submitted as its value

File: evaluation/diagramLogs.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                    txNumber = line[15:-1].split()[4]
+                    resultsSet.update({txNumber: (line[0:8])})
+                    #setFlag = True
+                if line.find("show") != -1:
```
Comment (WolfWindOW) [inline evaluation/diagramLogs.py:34]: If a line has the word "show" in it, it checks to see if there is a key with the same tx number in the map. If yes, add to the results

File: evaluation/diagramLogs.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    if resultsSet.get(line[15:-1].split()[4]) is not None:
+                        linkLines.append((resultsSet.get(line[15:-1].split()[4]), line[0:8]))
+                prev = line
+        resultsLink.append(linkLines)
+    results = []
+    #For each in resultsLink go through link lines with correct time
+    #round = 0
+    #For each log file
+    for list in resultsLink:
```
Comment (WolfWindOW) [inline evaluation/diagramLogs.py:43]: Add up all of the results, and output

File: evaluation/saturatedTransactions.py [hunk 6]
```diff
@@ -55,6 +56,8 @@ def get_avg_for(number_of_transactions: int, number_of_intersections: int, exper
         results = run_experiment(peers, number_of_transactions)
         throughputPer5[e] = results["throughput"]
         print("Cleaning up experiment {}".format(e))
+        os.echo("docker kill $(docker ps -q)")
```
Comment (WolfWindOW) [inline evaluation/saturatedTransactions.py:59]: Tries to clean up docker containers between runs

File: evaluation/saturatedTransactions.py [hunk 7]
```diff
@@ -116,10 +119,14 @@ def run_experiment(peers: dict, number_of_transactions: int):
 
 
 def check_from_peers(submitted, confirmed, peers, peerList):
-    intersectionA = peers[peerList[0]].inter
     for tx in submitted:
-        if intersectionA.get_tx(tx[1]) == tx[1].value:
```
Comment (WolfWindOW) [inline evaluation/saturatedTransactions.py:121]: If the transaction value was found, add it to confirmed

File: evaluation/saturatedTransactions.py [hunk 8]
```diff
@@ -116,10 +119,14 @@ def run_experiment(peers: dict, number_of_transactions: int):
 
 
 def check_from_peers(submitted, confirmed, peers, peerList):
-    intersectionA = peers[peerList[0]].inter
     for tx in submitted:
-        if intersectionA.get_tx(tx[1]) == tx[1].value:
-            confirmed[tx] = tx
+        for peer in peerList:
```
Comment (WolfWindOW) [inline evaluation/saturatedTransactions.py:123]: If the transaction value was found, that means it's been confirmed. Add it to confirmed.

File: evaluation/timingDiagram.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    avgs = get_avg_for(number_of_tx, experiment_duration_secs, measurement_interval_secs, number_of_intersections,
+                           experiments, committees)
+    print("Experiments for transaction amount {} ended".format(number_of_tx))
+    print("----------------------------------------------------------")
+    out.write("{s}, {w}\n".format(s=number_of_tx, w=avgs["throughput"]))
+    out.close()
+
+
+# Run each experiment and calc avgs
+def get_avg_for(number_of_transactions: int, experiment_duration_secs: int, measurement_interval_secs: int,
```
Comment (WolfWindOW) [inline evaluation/timingDiagram.py:47]: Doesn't actually get an average. Should probably refactor.

File: evaluation/timingDiagram.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        print("Setting up experiment {}".format(e))
+        peers = make_intersecting_committees_on_host(committees, number_of_intersections)
+        results = run_experiment(peers, experiment_duration_secs, measurement_interval_secs, number_of_transactions)
+        throughputPerE[e] = results["throughput"]
+        print("Cleaning up experiment {}".format(e))
+        del peers
+        gc.collect()
+    throughput = {}
+    # For each experiment
+    for e in range(experiments):
```
Comment (WolfWindOW) [inline evaluation/timingDiagram.py:60]: Basically just runs through all of the experiment's results, adds them together, and then outputs them in ascending order of key

File: evaluation/timingDiagram.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+# Gets the individual data for each amount of transactions sent
+# peers: dict of peers port number as key and SmartShardPeer as obj
+# experiment_duration_secs: how long to run experiment for
+# number_of_transactions: number of transactions per round (1 sec)
+def run_experiment(peers: dict, experiment_duration_secs: int, measurement_interval_secs: int,
+                   number_of_transactions: int):
+    print("Running", end='', flush=True)
+    committee_ids = [peers[p].committee_id_a() for p in peers]
+    committee_ids.extend([peers[p].committee_id_b() for p in peers])
+    committee_ids = list(dict.fromkeys(committee_ids))
```
Comment (WolfWindOW) [inline evaluation/timingDiagram.py:87]: Basically just creating two different ways to see what peer is in what quorum

File: evaluation/timingDiagram.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    peers_by_quorum = peersByQuorum(committee_ids, peers)
+    unsubmitted_tx_by_round = []  # list of transactions that should be submitted per round
+    txTimeSubAndConf = []
+    # Creates the amount of groups of tx equal to the amount of runs
+    for round in range(0, experiment_duration_secs):
```
Comment (WolfWindOW) [inline evaluation/timingDiagram.py:92]: Creates however many tx for each second in the experiment

File: evaluation/timingDiagram.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+def submitTxs(tuplesList):
+    # Submits the transaction, using the url as a key
+    requests.post(tuplesList[0], json=tuplesList[1].to_json())
+
+
+def setTime(txTime):
+    txTime.append(floor(time.time()))
+
+
+def create_txs(peers, committee_ids, number_of_transactions, round, txTimeSubAndConf):
```
Comment (WolfWindOW) [inline evaluation/timingDiagram.py:161]: Outputs a list of list of url, tx, time of creation tuples.

**Review by khood5 — APPROVED** [review]
Look ok to me

**Pull request merged**